src/streaming/controllers/BufferController.js

```diff
@@ -254,18 +254,27 @@ function BufferController(config) {
                 logger.warn('Quota exceeded, Critical Buffer: ' + criticalBufferLevel);
 
                 if (criticalBufferLevel > 0) {
-                    // recalculate buffer lengths to keep (bufferToKeep, bufferAheadToKeep, bufferTimeAtTopQuality) according to criticalBufferLevel
+                    // recalculate buffer lengths according to criticalBufferLevel
                     const bufferToKeep = Math.max(0.2 * criticalBufferLevel, 1);
                     const bufferAhead = criticalBufferLevel - bufferToKeep;
-                    const s = { streaming: { bufferToKeep: parseFloat(bufferToKeep.toFixed(5)),
-                                           bufferAheadToKeep: parseFloat(bufferAhead.toFixed(5))}};
+                    const bufferTimeAtTopQuality = Math.min(settings.get().streaming.bufferTimeAtTopQuality, bufferAhead * 0.9);
+                    const bufferTimeAtTopQualityLongForm = Math.min(settings.get().streaming.bufferTimeAtTopQualityLongForm, bufferAhead * 0.9);
+                    const s = {
+                        streaming: {
+                            bufferToKeep: parseFloat(bufferToKeep.toFixed(5)),
+                            bufferAheadToKeep: parseFloat(bufferAhead.toFixed(5)),
```

Comment:
I agree on your comment in the description:
If bufferAheadToKeep < bufferTimeAtTopQuality we request segments and remove them from the buffer. Same for bufferTimeAtTopQualityLongForm and stableBufferTime. The maximum buffer is limited by bufferTimeAtTopQuality and bufferTimeAtTopQualityLongform. Imo we can remove bufferAheadToKeep.

Reply:
OK, so we agree that buffer pruning process shall remove only part before currentTime (start = currentTime - buffertoKeep) ?